common/step_create_cdrom.go

```diff
@@ -167,12 +168,34 @@ var supportedCDISOCreationCommands []cdISOCreationCommand = []cdISOCreationComma
 	},
 }
 
+func isCygwinExecutable(path string) bool {
+	return runtime.GOOS == "windows" && strings.Contains(path, "\\usr\\bin\\")
```

Comment:
What could happen if you check only the OS here? 
Our pipeline run tests for Windows. You can add some tests for this with a toggle to run them only on Windows.

Reply:
we want to limit the path conversion step to cygwin/msys2 applications/tools. if the user has a native xorriso.exe file (or mkisofs, or any other of the supported tools to create an iso) we want to pass the native path (skipping the conversion step).

I will look at adding a windows test. Can you point me at one that runs in windows? I will need to have msys2 and install xorriso if they are not already there.

Reply:
Yes, any other path could be messed up.

I will try to do that unit test as suggested :-) But the `xorriso` command that needs to be tested is not native; xorriso has to be installed in msys2 (which is also a, hugely useful, 3rd party), which I will have to figure how to install in the packer CI.

Reply:
Never mind, it might be too much work. I think is good if you only add the template for testing this to your PR description. Other then that, the PR is good for me. I'll approve it!

Reply:
I will gladly do it. I just figured out [how to install/use msys2 in circle ci](https://app.circleci.com/pipelines/github/rgl/circleci-demo-windows/12/workflows/7872201b-6255-4842-a479-e2e99a94ef8c/jobs/12); it adds about 1m20s to the build time (hope that's not a problem!).

Though, I'm not yet sure how to modify [the packer ci pipeline](https://github.com/hashicorp/packer/blob/master/.circleci/config.yml) to run the tests; what do you think about adding a new `test-windows-msys2` job? similar to the existing `test-windows`, but it would run inside the msys2 environment (where it would have access to xorriso).

Reply:
I went ahead and implemented this in https://github.com/rgl/packer/commit/f592a39c6666095adb75863bf103555aa5ec57fa#diff-1d37e48f9ceff6d8030570cd36286a61 (please ignore the removed lines; they were only needed in order be able to run the packer CI in my circleci account).

You can see the results in https://app.circleci.com/pipelines/github/rgl/packer/9/workflows/9dc33fe7-81dd-4641-aaba-80316eeafe7e/jobs/33.

What do you think? Should that be added to this PR?

Please note that I still didn't made the actual unit test, I just proved that the packer CI can be changed to include msys2 tests.

Reply:
That's really nice @rgl! I like the new job, it will help us keep track of it, and also I think you can make sure to run only the test that is using the msys2 there, because the previous step is already running the rest of the unit tests. You can do that, for example, with `go test ./... -run=TestStepCreateCD_` to run all the tests in this step. 

About the 1m20s being add up, well I think we will only know how much that is a problem when we merge it. Maybe it will never be a problem and I think is worthy trying it out.

Reply:
About the test itself, you can actually run it for all the environments and create a structure which checks results. Like this one here: https://github.com/hashicorp/packer/blob/master/builder/vsphere/iso/step_create_test.go#L32 
Then you will make sure it works as expected for all environments and tests cases 👍   
You should add it to this PR, it will make sense.

Reply:
the `-run` seems pretty nice to use. instead of specifying the actual test name, how about using `-run msys2`? this way only tests that have `msys2` in their name would be executed.

or maybe there's a way to use build constraints/flags/tags? that is, put these msys2 in separate files with the suffix `_msys2_test.go`, have a file build constraint alike `// +build msys2`, then run the tests with `go test ... -tags msys2`.

I didn't see the `-coverprofile=coverage.txt -covermode=atomic` arguments in your comment; should they still be used?

I didn't understood your suggestion about the test itself; currently there's only a single test and as such I see no need to have a more complex test structure. Or did you meant something else?

Reply:
When running the test `TestStepCreateCD` would the `retrieveCDISOCreationCommand` be covered as well?  If yes, then running it will also be nice for this case, cuz is a more "integrated" test with the rest of the step.

I'll let you decide what you think is best to run the specific tests based on the answer to my question above. I like the tags option as well because the file will be excluded from the other builds.

Reply:
Regarding of how I suggested to write the test, I thought of something like 

```
func TestStepCreateCD_retrieveCDISOCreationCommand_msys2(t *testing.T) {
	tc := []struct{
		name                    string
		goos                    string
		isCygpath               bool
		source                  string
		dest                    string
		pathSuffix              string
		pathLength              int
		expectedSourcePath      string
		expectedDestinationPath string
	} {
		{
			name:                    "Windows with cygpath",
			goos:                    "windows",
			isCygpath:               true,
			source:                  "C:\\Windows\\Temp\\test-cd",
			dest:                    "C:\\Windows\\Temp\\test-cd.iso",
			pathSuffix:              "\\usr\\bin\\xorriso.exe",
			pathLength:              2,
			expectedSourcePath:      "/c/Windows/Temp/test-cd",
			expectedDestinationPath: "/c/Windows/Temp/test-cd.iso",
		},
		{
			name:                    "Linux",
			goos:                    "linux",
			isCygpath:               false,
			source:                  "/usr/src",
			dest:                    "/usr/dst",
			pathSuffix:              "/usr/path/suffix",
			pathLength:              2,
			expectedSourcePath:      "/usr/src",
			expectedDestinationPath: "/usr/dst",
		},
	}

	for _, c := range tc {
		t.Run(c.name, func(t *testing.T) {
			if runtime.GOOS != c.goos {
				t.Skip(fmt.Sprintf("test only applies to %s", c.goos))
			}
			if c.isCygpath {
				_, err := exec.LookPath("cygpath")
				if err != nil {
					t.Skip("test only applies to windows msys2")
				}
			}
			command, err := retrieveCDISOCreationCommand(
				"test",
				c.source,
				c.dest)
			if err != nil {
				t.Fatalf("should not fail")
			}
			if !strings.HasSuffix(command.Path, c.pathSuffix) {
				t.Fatalf("expected a command path but got %s", command.Path)
			}
			if len(command.Args) < c.pathLength {
				t.Fatalf("expected a command args length >= 2 but got %v", command.Args)
			}
			sourcePath := command.Args[len(command.Args)-1]
			if c.expectedSourcePath != sourcePath {
				t.Fatalf("expected the source path to be %s but got %s", c.expectedSourcePath, sourcePath)
			}
			destinationPath := command.Args[len(command.Args)-2]
			if c.expectedDestinationPath != destinationPath {
				t.Fatalf("expected the destination path to be %s but got %s", c.expectedDestinationPath, destinationPath)
			}
		})
	}
}
```

I don't what valid values would be for the other OSes and cases, but with that structure will be possible to test multiple possibilities. Since you know better what's happening in `retrieveCDISOCreationCommand` I'll let you decide that too.